[FIXED JENKINS-30559] Dropdown list to select commit info for notifications

ahippo opened the pull request:
Convert existing checkbox "Show Commit List with Title and Authors"
into a dropdown list with the following options:
 - nothing about commits
 - commit list with authors only
 - commit list with authors and titles

This list can be easily extended with new options.

ahippo commented:
@samrocketman, have you had a chance to review the pull request?
I'm sorry to disturb you directly but the [aforementioned](https://github.com/jenkinsci/slack-plugin/pull/133#issuecomment-141883515) ["Pull Request to Repositories"](https://wiki.jenkins-ci.org/display/JENKINS/Pull+Request+to+Repositories) document suggests to go ask for commit access to the repository "after a week or so".
So, I decided to first ping you in case you reviewed it already.

samrocketman commented:
I haven't reviewed it, yet.  My computer has been broken though in general I also leave it open for review for any community members at least a week.

samrocketman commented:
cc @cheries @FaizVisram @tterrag1098 @mattwilliams-stfc @schisamo @mboersma @bigwheel @chEbba @philowest

What are your thoughts on this behavior change.  Since you've recently opened pull requests do you mind reviewing this one?  I'll do my best to review it as well.  Thanks.

I'll address this once a few sets of eyeballs peruse it.

tterrag1098 commented:
Well I'll be honest, I don't know much about coding for jenkins, but the one thing I notice is that the enum is a bit verbose, it could just take booleans for the different options, and the override could be reserved for more complicated behavior. Also can the text in the options be localized? Seems hardcoded to me.

samrocketman commented:
@tterrag1098 thanks for your review.  Regarding your questions can you comment on specific lines so that your feedback is more pointed?

tterrag1098 commented on src/main/java/jenkins/plugins/slack/CommitInfoChoice.java:
```diff
@@ -0,0 +1,61 @@
+        this.displayName = displayName;
+    }
+
+    public abstract boolean showAuthor();
```
Add default impls for these two methods that return a boolean, which is passed to the constructor. This can eliminate about 20 LoC.

ahippo replied:
Yeah, makes sense.
Thank you for the suggestion!

tterrag1098 commented on src/main/java/jenkins/plugins/slack/CommitInfoChoice.java:
```diff
@@ -0,0 +1,61 @@
+package jenkins.plugins.slack;
+
+public enum CommitInfoChoice {
+    NONE("nothing about commits") {
```
Are these strings localizeable?

ahippo replied:
Frankly speaking, I know nothing about localization in Jenkins.
According to [Internationalization page](https://wiki.jenkins-ci.org/display/JENKINS/Internationalization), there must be a *.properties file in the source tree, so I doubt this plugin currently has support for localization.

While it's possible to add this support, I think the change deserves a separate pull request.

I can convert them to String constants though if it helps in the future to add l10n support.

tterrag1098 replied:
That's fine, like I said I'm unfamiliar. Probably ok to leave it as hardcoded for now.

ahippo commented:
> Well I'll be honest, I don't know much about coding for jenkins, but the one thing I notice is that the enum is a bit verbose, it could just take booleans for the different options, and the override could be reserved for more complicated behavior. Also can the text in the options be localized? Seems hardcoded to me.

Thank you for your review!
I'll make the change for the enum and move the strings to String constants.

samrocketman commented:
@ahippo are there any additional changes that need to be made to address @tterrag1098's feedback?  If not then I will change the status of this to `candidate to merge`.

ahippo commented:
> @ahippo are there any additional changes that need to be made to address @tterrag1098's feedback? If not then I will change the status of this to candidate to merge.

Yeah, I'll make a couple of changes to address his comments.
Hopefully, I'll push them here by the weekend.

samrocketman commented:
Okay, I'll wait until you make the changes then so it's not merged pre-maturely.

ahippo commented:
> Okay, I'll wait until you make the changes then so it's not merged prematurely.

Thank you!
I'll let you know as soon as I make my changes.

ahippo commented:
@samrocketman, I've just pushed the changes I wanted to make.

I addressed the first @tterrag1098's comment only.
I tried to address the second one too but it started to cause some problems, so I dropped it for now given that it's not clear whether it will help with localization or not.

ahippo commented:
@samrocketman, thank you for merging!

samrocketman commented:
Thanks for contributing.  If you'd like, I'm slack plugin users would appreciate #106 being addressed.  I'll provide any support around it I can.

samrocketman commented:
FYI you can quickly and easily test migration by spinning up a pre-configured instance of Jenkins.  See https://github.com/samrocketman/jenkins-bootstrap-slack

ahippo commented:
> Thanks for contributing. If you'd like, I'm sure slack plugin users would appreciate #106 being addressed. I'll provide any support around it I can.

Thank you for pointing this out!
However, I'm not sure I have enough time and understanding of Jenkins/Plugin internals to address the issue at this moment.
So, most likely, I won't be able to take a look in the near future.

> FYI you can quickly and easily test migration by spinning up a pre-configured instance of Jenkins. See https://github.com/samrocketman/jenkins-bootstrap-slack

Oh,  thank you, good to know!
I actually had to do this manually.
